### ISSUE TITLE
Combining nextest, llvm-cov and delayed execution

### ISSUE THREAD

**Issue by MiniaczQ** [nextest-rs/nextest#1036]
Hello, I'm investing how I can create an optimal integration test environment and that includes code coverage and junit compatible test logs, while efficiently using build cache.
I believe cargo-llvm-cov combined with nextest it possible, but I'm having some trouble executing it and I was hoping someone could help me out here.
```docker-compose
FROM rust:1.73-bookworm as build
WORKDIR /app/

RUN rustup component add llvm-tools-preview
RUN curl -L --proto '=https' --tlsv1.2 -sSf https://raw.githubusercontent.com/cargo-bins/cargo-binstall/main/install-from-binstall-release.sh | bash
RUN cargo binstall cargo-nextest --secure -y
RUN cargo binstall cargo-llvm-cov --secure -y

COPY . .
# Skipped cache mounting, that works fine
# Prepare build flags for coverage, save it in a file to execute in each RUN statement
RUN echo $(cargo llvm-cov show-env --export-prefix) >> ./envvars.sh 
RUN chmod +x ./envvars.sh
# Build all dependencies with the coverage flags and nextest
RUN ./envvars.sh && cargo nextest run --no-run

FROM build as unit-test
WORKDIR /app/
# Everything is already built, just export the correct tests to an archive
RUN ./envvars.sh && cargo nextest archive --archive-file test.tar.zst --lib --bins
ENTRYPOINT ./tests/run-report.sh
```
The `run-report.sh`
```sh
set -e
# Run tests from the archive, it doesn't make sense to still add build flags here
cargo nextest run --archive-file test.tar.zst
# Use generated profraws for report generation
cargo llvm-cov report --html
```

The issue I'm running into is that the profraws aren't being generated and I believe this has to do with the nextest building process.
Any ideas?

**Comment by MiniaczQ** [comment]
Don't get confused by the example, I named the target `unit-test`, but I just treat it as special case integration tests with no external services.

**Comment by taiki-e** [comment]
> ```
> echo $(cargo llvm-cov show-env --export-prefix) >> ./envvars.sh 
> ```

IIRC, this `echo` converts the output from show-env to a space-separated single line, which will result in environment variables being set incorrectly. I think you have to use `cargo llvm-cov show-env --export-prefix >> ./envvars.sh`

**Comment by MiniaczQ** [comment]
You're correct, I tried the second one and it suffers from the same issue

**Comment by MiniaczQ** [comment]
`source` doesn't exist in whichever shell RUN is using, so I left it early, but `.` seems to at least go through, I'm checking whether it exports properly

**Comment by MiniaczQ** [comment]
Export works, no profraws though :/

**Comment by taiki-e** [comment]
> ```
> set -e
> # Run tests from the archive
> ./envvars.sh && cargo nextest run --archive-file test.tar.zst
> # Use generated profraws for report generation
> cargo llvm-cov report --html
> ```

- --archive-file is not supported yet. Try https://github.com/taiki-e/cargo-llvm-cov/pull/266 with [my patch](https://github.com/taiki-e/cargo-llvm-cov/pull/266#issuecomment-1527785686).
- `cargo llvm-cov report` also needs to reference the environment variables from show-env.

**Comment by taiki-e** [comment]
> * --archive-file is not supported yet. Try [Use extract-to if using --archive-file with nextest taiki-e/cargo-llvm-cov#266](https://github.com/taiki-e/cargo-llvm-cov/pull/266) with [my patch](https://github.com/taiki-e/cargo-llvm-cov/pull/266#issuecomment-1527785686).

Oh, that's about the builtin nextest subcommand, so that may not work for show-env as-is.

**Comment by MiniaczQ** [comment]
> * --archive-file is not supported yet. Try [Use extract-to if using --archive-file with nextest taiki-e/cargo-llvm-cov#266](https://github.com/taiki-e/cargo-llvm-cov/pull/266) with [my patch](https://github.com/taiki-e/cargo-llvm-cov/pull/266#issuecomment-1527785686).

I'm not sure I understand the issue, profraws should be generated during compilation and they don't get packaged with `nextest archive`? so tests don't generate anything?

> * `cargo llvm-cov report` also needs to reference the environment variables from show-env.

Added it, but I checked and there are just no .profraw's after the test run

**Comment by MiniaczQ** [comment]
Would moving the file by hand potentially fix it?

**Comment by taiki-e** [comment]
> profraws should be generated during compilation

No, binary (or cdylib) will generate *.profraw file on exit. See also https://github.com/taiki-e/cargo-llvm-cov/issues/320#issuecomment-1751966574.

> no profraws

Does this mean it cannot be found using `find` or `fd -I`? Or does it mean that `cargo llvm-cov report` cannot find it?

In the latter case, I guess the problem is that the target directory used by --archive-file is different from cargo's default, just like the problem that occurred in the linked PR, so I think it can be avoided by adjusting `CARGO_LLVM_COV_TARGET_DIR` (maybe adding `/target` at last) before calling `cargo llvm-cov report`.

**Comment by MiniaczQ** [comment]
Docker desktop highlights changed and new files in a container, they just don't exist.
There are only my tmp files for tests and this:
![image](https://github.com/nextest-rs/nextest/assets/13040204/2387a2d8-8cfa-4f92-8bf5-760426412609)

**Comment by MiniaczQ** [comment]
Here, I made a repro repository, the problem is in the issue
https://github.com/MiniaczQ/nextext-llvm-cov-repro

**Comment by MiniaczQ** [comment]
Solved thanks to a friend
